### ISSUE TITLE
[Extract PNG  From PDF] the extracted picture has some pixels with different color from what we see in PDF

### ISSUE THREAD

**Issue by YongxinLai** [pymupdf/PyMuPDF-Utilities#6]
The picture in PDF has alpha value. So when I extract png from PDF, I get the xref first and then set smask samples value as alpha. It works. However, I found there are several pixel which has different value and color from what we see in PDF. Do you have any ideas or suggestions? Thanks in advance.

**Comment by YongxinLai** [comment]
In addition, if I don’t set the alpha, the wrong color  also appear in the same position.

**Comment by JorjMcKie** [comment]
> I found there are several pixel which has different value and color from what we see in PDF

Hard to tell without seeing an example PDF and the code snippet extracting the PNG.

In general, transparent images in PDF are stored as separate sub images: one image contains the color pixels, e.g. (R, G, B) for an RGB image, and the other the alphas (transparency in formation).
If you extract such an image and make a pixmap of it, it will contain the 4-tuple (R, G, B, A) for each pixel. All these 4 values are integers from 0 to 255. In case of the alpha value A, this is interpreted as transparency, e.g. 128 => 128/255 ~ 50%.
The ``setAlpha`` method changes all those alpha values in a pixmap - based on an array of integers. If all values are set to 255, the image will no longer be transparent. If 0, it will be fully transparent and you won't see anything any more, etc.

**Comment by JorjMcKie** [comment]
Closing for lack of reaction.

**Comment by meruiden** [comment]
I am having the same issue. however some images get fixed when i wrap it with ```fitz.Pixmap(fitz.csRGB, fitz.Pixmap(pix1))``` right before setting the alpha. but when I do that some other images get these weird artifacts again.

this is how the pdf looks:
![pdf](https://s3.csteenhuis.com/sharex/DMYjrKdDuX2s6RY.png)

this is without the conversion before applying the alpha:
![without](https://s3.csteenhuis.com/sharex/9l7lKrQzKf8x5WB.png)

this is with the ```fitz.csRGB```conversion
![with](https://s3.csteenhuis.com/sharex/4FfCod6YoqGyoYp.png)

you can reproduce this by changing the recoverpix function.

this is the code I used (its basically a simple version of recoverpix)

````
def recoverpix(doc, item):
    x = item[0]  # xref of PDF image
    s = item[1]  # xref of its /SMask

    pix1 = fitz.Pixmap(doc, x)
    if s == 0:
        return fitz.Pixmap(fitz.csRGB, pix1)

    pix2 = fitz.Pixmap(doc, s)

    #pix3 = fitz.Pixmap(fitz.csRGB, fitz.Pixmap(pix1)) # with conversion
    pix3 = fitz.Pixmap(pix1) # without conversion

    pix3.setAlpha(pix2.samples)

    return fitz.Pixmap(fitz.csRGB, pix3)
```

any idea @JorjMcKie ?

**Comment by JorjMcKie** [comment]
Ok, thanks for the hint concerning wand (never heard of it before).
I'll have a look into it.

**Comment by meruiden** [comment]
> Ok, thanks for the hint concerning wand (never heard of it before).
> I'll have a look into it.

No problem. I made some small changes to the code tog et it a little cleaner but the functionality is still the same. (edited the original post. I hope this helps finding the issue with the random pixels

**Comment by JorjMcKie** [comment]
@meruiden 
My approach to recover the original, transparency-loaden image simply is too naive.
MuPDF actually does provide the necessary functionality for recovering that original. I erroneously thought I can take this type of shortcut.

What I will do, is reworking the ``doc.extractImage()`` method such that it uses MuPDF more consequently and either **_hide_** any smask xrefs (e.g. by always set them to 0) in the response dictionary, and / or change the example scripts containing that cited type of code.

**Comment by JorjMcKie** [comment]
@meruiden - could you please send me one of those problem examples?
Made a few changes and would like to test ...

**Comment by JorjMcKie** [comment]
@meruiden - in the meantime I have changed my position on this: I gave up resolving issues around transparent images using MuPDF. There simply is no reliable way to do it (or I am too dumb figuring it out).
Their own code also has defects around the same topic. If you `page.apply_redactions()` overlapping transarent images, those images will be updated ignoring their /SMask. Similarly if you extract page text to HTML, transparent images are not processed correctly, etc.

Instead I am using PIL/Pillow. The image extraction scripts in the examples directory already are updated accordingly. I hope they are worry-free to use now.
